### DISCUSSION TITLE
YeeLight with WROOM32D ESP32

### DISCUSSION THREAD

**Discussion by elajoie** [arendst/Tasmota#10762] [Q&A] [answered]
Hi all,

Has anyone gotten the YeeLight images and working?
https://templates.blakadder.com/yeelight_YLDL01YL.html

I have tried different methods via factory image and Tasmota via esptool for example:

`esptool.py --chip esp32 --port /dev/ttyUSB0 --baud 921600 --before default_reset --after hard_reset write_flash -z --flash_mode dout --flash_freq 40m --flash_size detect 0x1000 bootloader_dout_40m.bin 0x8000 partitions.bin 0xe000 boot_app0.bin 0x10000 tasmota32solo1.bin`

Every time I try it ends up in a reboot loop. The best I could get was a factory image then write in the solo1 image into 0x10000.

I also tied both the tasmota32 and tasmota32solo1 build options but both result in continual reboots as well.

Thanks,
Eric

**Comment by Jason2866** [comment]
My automated builds are NOT tested! Try to flash the tasmota32solo1 to a ESP32 dev board.
The solo1 build runs on a usual board too. If it doesnt start the build is not working
Btw. how have you downloaded the firmware? Doing with browser "save as" will give a corrupt file!

**Comment by elajoie** [comment]
I did do the mistake you mentioned early on until I used the download button.  Noticed this when I tried to look into the partitions.bin and notice javascript xml content.

Here is my working files files from your custom build plus the factory image:
```
-rw-r--r--. 1 elajoie elajoie    8192 Jan 30 16:30 boot_app0.bin
-rw-r--r--. 1 elajoie elajoie   15872 Jan 30 16:30 bootloader_dout_40m.bin
-rw-r--r--. 1 elajoie elajoie 4194304 Jul 27  2020 factory_WROOM-32.bin
-rw-r--r--. 1 elajoie elajoie    3072 Jan 30 16:30 partitions.bin
-rw-r--r--. 1 elajoie elajoie 1414800 Jan 30 16:30 tasmota32solo1.bin
```

Martin seems to have it working but my method was exactly the same and only difference is he used an image from January 6ish I am guessing.  https://photos.app.goo.gl/T3F4aaXuYgvFXRmi7

I will order a couple wroom32d dev boards to see if there is a difference but I see the same result with two of these lights.

Do you have any info on what you set in your override file when making the build?

Thanks for the help!

**Comment by Jason2866** [comment, marked as answer]
Around this date i used this version
`framework-arduinoespressif32 @ https://github.com/Jason2866/esp32-arduino-lib-builder/raw/framework-arduinoespressif32/framework-arduinoespressif32-release_v3.3-solo1-bd65eb8d1.tar.gz
`

**Reply by elajoie** [reply]
> Around this date i used this version
> `framework-arduinoespressif32 @ https://github.com/Jason2866/esp32-arduino-lib-builder/raw/framework-arduinoespressif32/framework-arduinoespressif32-release_v3.3-solo1-bd65eb8d1.tar.gz `

Yup, thanks Jason this made it flawless.

**Comment by avoll94** [comment]
Hi, 

I have a similar problem with the YeeLight product.
I have managed to flash the esp32 as described in the instructions, but am stuck in a boot loop after rebooting too.

`
Rebooting...
ets Jun  8 2016 00:22:57

rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DOUT, clock div:2
load:0x3fff0018,len:4
load:0x3fff001c,len:1044
load:0x40078000,len:8896
load:0x40080400,len:5828
entry 0x400806ac

00:00:00.005 UFS: FlashFS mounted with 52 kB free
00:00:00.018 CFG: Loaded from File, Count 1
00:00:00.021 FRC: Some settings have been reset (68)
E (446) system_api: Base MAC address from BLK0 of EFUSE CRC error, efuse_crc = 0x8b; calc_crc = 0x76
abort() was called at PC 0x401528e3 on core 0

ELF file SHA256: 0000000000000000

Backtrace: 0x40090197:0x3ffcf410 0x400903fd:0x3ffcf430 0x401528e3:0x3ffcf450 0x40152939:0x3ffcf490 0x401046e9:0x3ffcf4c0 0x400da7d1:0x3ffcf510 0x400f93b9:0x3ffcf5d0 0x4012f702:0x3ffcf620 0x400913a6:0x3ffcf640

Rebooting...
`
Perhaps the problem is similar and can be treated.

Thanks,
Andre

**Comment by Jason2866** [comment]
@avoll94  can you try the framework version i mentioned (older one from January)?

**Reply by avoll94** [reply]
I am missing information on how to create a binary file from the framework. Last time I took the finished bin and burned it on the esp. Maybe someone can send me the bin or show me how to generate it?

**Reply by ascillato** [reply]
https://tasmota.github.io/docs/Compile-your-build/

**Comment by Jason2866** [comment]
THX for feedback! PR #10778 to use (older) know working version for ESP32solo1

**Comment by Kugeleis** [comment]
The ESP WROOM 32D in the meteorite should also eat normal tamota32 builds since it is a dual core.
Are there experiences with recent builds?

**Reply by sfromis** [reply]
The one in my Metorite is single core, and works fine on current tasmota32solo1.bin. Do not believe what is printed on the can, Xiaomi/Yeelight seems to deviate from the most common naming.

**Reply by Jason2866** [reply]
Not only the printing on the can is cheating, Xiaomi / Yeelight ESP32-solo1 are special OEM variants which do NOT work with the provided Espressif IDF framework. To get it working a patch of the IDF code is neccessary. Afaik Tasmota is the only Open Source firmware running on Xiaomi Yeelight devices.

**Reply by Kugeleis** [reply]
You're right. The can print of my lamp is misleading. Only the solo version works. But at least Bluetooth and the hall sensor are in and working.

**Reply by Jason2866** [reply]
Enabling Bluetooth with a one core ESP32 can be too much. If you experience bad performance or WDTs that is the reason for.

**Reply by sfromis** [reply]
... and using the Hall Effect sensor on a meteorite may not be a very practical use case ...

**Comment by opetergmxnet** [comment]
Hi, is it true that I do have to use ESPHome to bring Tasmota to this device?

 I can not flash Tasmota directly with: tasmota.github.io/install/  ?

**Reply by Jason2866** [reply]
Just use the Tasmota Web Installer and select the solo1 build. What is EspHome ;-)

**Reply by opetergmxnet** [reply]
Ha! Confirmed! Web Installer and pulling low brings me to bootmode and... voilà! Thanks very much.

**Comment by MiGeppert** [comment]
Hi guys,
I tried to find some information about that "hall switch"!
What is this for?

And by the way - yesterday I finally found the time to flash Tasmota on my Meterorite/Crystal.
Is there a possibility to dim the main light "further down"? Depending on the color temperature, the light is already off at/below 40-50% PWM.
Any ideas or any ways to go down to 10% or lower?

Thanks in advance!

Best regards
Cobraball

**Reply by sfromis** [reply]
Hall Effect is not a switch, but a sensor, which you could use to detect a magnet in relatively close proximity. Not a great use case in the pendent light, but you could get varying numbers by repositioning a strong magnet on top of the lamp, near the center, and use that as a basis for doing something.

To make the dimming slider more useful, I'm using `DimmerRange 40,100` to have the full slider range useful for the lamp, but indeed not really low brightess. As the bad color temperature variations alredy make the lower power part a lot less useful, the lower 40 is probably too low.

As the template page https://templates.blakadder.com/yeelight_YLDL01YL explains, gpio23 is for the night light feature of the lamp, which goes way lower than the regular lamp brightness. To use this, you'd need some Berry code for this 6th PWM channel, and maybe then add another slider to the web UI.

Some time ago, I was thinking about integrating the night light into the normal brighness slider with some Berry code, but did not work on it, as my usage does not include the night light.

**Reply by sfromis** [reply]
Here's a bit of sample Berry code for having a simplistic slider feature added, to control night light.
```
class Slider

  var id,min,max,value,action

  def init(id,min,max,startvalue,action)
    self.id = id
    self.min = min
    self.max = max
    self.value = startvalue
    self.action = action == nil ? /value->print(f'slider {self.id} {self.value}') : action
    var globalname = 'Slider_instance_' + self.id
    tasmota.remove_driver(global.member(globalname))
    tasmota.add_driver(self)
    global.setmember(globalname, self)
    var f = self.action f(self.id, self.value)
  end

  def web_add_main_button()
    import webserver
    var onchange=f'la("&{self.id}="+value)'
    webserver.content_send(f'<input type="range" min={self.min} max={self.max} value={self.value} {onchange=} id="{self.id}" class="slider">') 
  end

  def web_sensor()
    import webserver
    if webserver.has_arg(self.id)
      self.value = int(webserver.arg(webserver.arg_name(1)))
      var f = self.action f(self.id, self.value)
    end
  end

end

slider = Slider('nightlight',0,1023,0, /id,level->gpio.set_pwm(23, level))
```
The `Slider` class is a generic one for any slider, while the last line is what links the slider to the PWM on gpio 23. A more complete integration would coordinate with what the "regular lamp part" is doing, including the "master" power on/off.

**Reply by MiGeppert** [reply]
Hi sfromis, 
thank you for your message, I will have a look to all these information!
But one info is missing: where is this hall switch connected to?
And I will check out the night light and see, if this is something I can use.

Best regards
Cobraball

**Reply by sfromis** [reply]
It is no switch, but the usual Hall Effect *sensor* in original ESP32. In Tasmota, you can configure gpio36,39 as HallEffect 1,2. After the restart, you see a number on the Tasmota front page, and available all places where the usual sensor payload is.

**Reply by MiGeppert** [reply]
Hi again,

I was not aware of the fact, that (a) hall _sensor(s)_ is/are integraded in the ESP32.
As I am familiar with hall _switches_, I thought one or more of them could be implemented in the lamp to e.g. switch on some special "things".
Anyway - thanks again for your information.
Now I have to find some time to test all that stuff.
I also have running the berry script on the lamp, the one that's described on the page you metioned above.

I'm sure there is also a way to implement the night lamp slider/function via tasmota template - I need to read about that what I have to do for this...

Best regards
Cobraball

**Reply by sfromis** [reply]
> I'm sure there is also a way to implement the night lamp slider/function via tasmota template - I need to read about that what I have to do for this...

No. Otherwise I'd have suggested that already. Closest is that you can config it as PWM index 6, and then use the command `PWM6` to control it.

What could be doable would be what I mentioned, having some Berry code sitting on top of what the slider is producing, and then control the nightlight PWM where you think it fits in, maybe on brightness levels lower than what the "normal" lamp can provide.

If you don't care about butchering the RGB part, it would be doable to config those 3 PWM channels elsewhere, with one of them on the nightlight pin. Then you could use the RGB controls for the nightlight. A bit of a mess, but the world is your oyster.
